backtrace-library/src/main/java/backtraceio/library/metrics/BacktraceMetrics.java

```diff
@@ -0,0 +1,309 @@
+     * Maximum number of events in store. If number of events in store hit the limit
+     * BacktraceMetrics instance will send data to Backtrace.
+     */
+    private int maximumEvents = 350;
+
+    /**
+     * Name of the summed event that will be generated on app startup
+     */
+    private final String startupSummedEventName = "Application Launches";
+
+    /**
+     * Maximum number of attempts
+     */
+    public final static int maxNumberOfAttempts = 3;
+
+    /**
+     * Default time between retries in seconds
+     */
+    public final static int defaultTimeBetweenRetriesMillis = 10000;
+
+    /**
+     * Maximum time between requests in minutes
+     */
+    public final static int maxTimeBetweenRetriesMillis = 5 * 60 * 1000;
+
+    /**
+     * Default submission url
+     */
+    public final static String defaultBaseUrl = "https://events.backtrace.io/api";
+
+    /**
+     * User custom request method
+     */
+    private RequestHandler requestHandler = null;
+
+    /**
+     * Unique Events handler
+     */
+    public UniqueEventsHandler uniqueEventsHandler;
+
+    /**
+     * Summed Events handler
+     */
+    public SummedEventsHandler summedEventsHandler;
+
+    /**
+     * Custom attributes provided by the user to BacktraceBase
+     */
+    protected Map<String, Object> customAttributes;
+
+    /**
+     * The application context
+     */
+    protected Context context;
+
+    /**
+     * Backtrace metrics settings
+     */
+    protected BacktraceMetricsSettings settings;
+
+    /**
+     * Create new Backtrace metrics instance
+     * @param context                   Application context
+     * @param customAttributes          Backtrace client
+     * @param settings
+     */
+    public BacktraceMetrics(Context context, Map<String, Object> customAttributes, BacktraceMetricsSettings settings) {
+        this.context = context;
+        this.customAttributes = customAttributes;
+        this.settings = settings;
+    }
+
+    @Override
+    public void startMetricsEventHandlers(Api backtraceApi) {
+        uniqueEventsHandler = backtraceApi.enableUniqueEvents(context, customAttributes, settings);
+        summedEventsHandler = backtraceApi.enableSummedEvents(context, customAttributes, settings);
+    }
+
+    void setStartupUniqueEventName(String StartupUniqueEventName) {
+        this.startupUniqueEventName = StartupUniqueEventName;
+    }
+
+    String getStartupUniqueEventName() {
+        return this.startupUniqueEventName;
+    }
+
+    void setMaximumEvents(int MaximumEvents) {
+        this.maximumEvents = MaximumEvents;
+    }
+
+    int getMaximumEvents() {
+        return this.maximumEvents;
+    }
+
+    public String getBaseUrl() {
+        return this.settings.getBaseUrl();
+    }
+
+    @SuppressWarnings("unchecked, rawtypes")
+    public ConcurrentLinkedDeque<UniqueEvent> getUniqueEvents() {
+        for (Event event : uniqueEventsHandler.events) {
+            if (!(event instanceof UniqueEvent)) {
```

Comment:
why do we need to check object type? isn't it weird that if we fail with object type comparison we're returning an empty collection and the user will think we don't have events in the queue (but we will actually)?

Reply:
Why we don't use generic parameters instead?

Reply:
We can. I thought it might add a little more type safety to at least ask for Event types specifically but maybe it doesn't matter @konraddysput .